### PR TITLE
[WIP] AppSync - type-safe Velocity Template and GraphQL schemas

### PR DESCRIPTION
**Pull request by sam-goodwin**
This is a Punchcard abstraction for AWS AppSync that unifies GraphQL, Velocity Templates and AppSync Resolvers into a single, type-safe environment. Developers write simple TypeScript code which then generates the GraphQL Schema, VTL templates and CDK configuration for an AWS AppSync API.

What I really love about this change is that it lets Punchcard abstract over AppSync's managed VTL execution engine. Effective use of these templates and resolvers can rapidly increase the reliability of a service - a developer has even less to manage than lambda.

Punchcard also brings type-safety to an otherwise annoying programming environment - VTL. The VTL DSL makes use of TypeScript's compiler to statically check that an API definition is sound.

https://github.com/punchcard/punchcard/blob/f3d6d8208208dc4b0aad84f0519ba18d26f96419/examples/lib/graphql.ts#L9-L188

### CODE REVIEW

File: packages/punchcard/lib/appsync/api.ts [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+    if (props.query !== undefined) {
+      this.QueryFQN = props.query;
+      this.QueryType = this.Types[props.query!].type as any;
+    }
+    if (props.mutation !== undefined) {
+      this.MutationFQN = props.mutation;
+      this.MutationType = this.Types[props.mutation!].type as any;
+    }
+
+    this.resource = Build.concat(
```
Comment (sam-goodwin) [inline packages/punchcard/lib/appsync/api.ts:71]: This is the magic - it interprets the Velocity template "programs" and synthesizes a GraphQL Schema and Resolver Pipelines.

File: examples/lib/graphql.ts [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    // dummy lambda function implemementation
+    return tags.map(tag => new Post({
+      id: 'id',
+      title: 'title',
+      content: `tag: ${tag}`,
+      tags
+    }));
+  });
+
+  // export thee API implementations
```
Comment (harrysolovay) [inline examples/lib/graphql.ts:144]: `thee` > `the`

File: packages/@punchcard/shape/lib/function.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  public readonly Kind: 'functionShape' = 'functionShape';
+  public readonly FQN: string;
+
+  constructor(
+    public readonly args: Args,
+    public readonly returns: Returns
+  ) {
+    super();
+    // compute the FQN
+    this.FQN = `(${Object.entries(args).map(([k, v]) => `${k}: ${v.FQN}`).join(',')}) => ${returns.FQN}`;
```
Comment (harrysolovay) [inline packages/@punchcard/shape/lib/function.ts:22]: In the next version of shapes, can the FQN be a trait? Not all shapes use cases will require an FQN... and although it's nice that the user doesn't need to assign one themselves, it doesn't feel to me like it should be built in. I'm wondering what you think.

```ts
import * as s from "shapes";

class MyRecord extends s.createVRecord({}) {}

class MyFQRecord1 extends MyRecord.with({ fqn: "first" }) {}
class MyFQRecord2 extends MyFQRecord1.with({ fqn: "second" }) {} // overrides "first"
```

Semi-related: when you extend a record shape, does the result carry the same FQN? Glancing through [the source](https://github.com/punchcard/punchcard/blob/master/packages/%40punchcard/shape/lib/record.ts#L251).

File: packages/punchcard/lib/appsync/fragment.ts [hunk 4]
```diff
@@ -0,0 +1,195 @@
+import { RecordShape, Shape, ShapeGuards } from '@punchcard/shape';
+import { TypeSpec, TypeSystem } from './type-system';
+
+export class ApiFragment<I extends TypeSystem = any> {
+  public static join<F1 extends ApiFragment>(
```
Comment (harrysolovay) [inline packages/punchcard/lib/appsync/fragment.ts:5]: Ughhh... these overloads are bumming me out. Hopefully [this](https://github.com/microsoft/TypeScript/issues/5453) happens soon :/

File: packages/punchcard/lib/appsync/trait.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        resolvers
+      }
+    } as any);
+  }
+}
+
+/**
+ * Implementation of the field resolvers in a Trait.
+ */
+export type TraitImpl<T extends RecordShape, F extends RecordMembers> = {
```
Comment (harrysolovay) [inline packages/punchcard/lib/appsync/trait.ts:73]: This is starting to feel quite complex. What other approaches have you played with for interconnecting the types from  traits and shapes?

File: packages/punchcard/lib/appsync/trait.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+  [f in keyof F]: F[f] extends FunctionShape<infer Args, infer Returns> ?
+    // if it's a Function type, expect a function taking those args and returning an object
+    (args: {
+      [arg in keyof Args]: VObject.Of<Args[arg]>;
+    }, self: VObject.Of<T>) => VTL<VObject.Of<Returns>> :
+    // no args if it is not a Function type
+    (self: VObject.Of<T>) => VTL<VObject.Of<F[f]>>
+} & ThisType<VObject.Of<T>> // make the value of self also available as `this`. (optional syntactic sugar).
```
Comment (harrysolovay) [inline packages/punchcard/lib/appsync/trait.ts:81]: Oh yes.

File: packages/punchcard/lib/appsync/type-system.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+/**
+ * A map from a type's FQN to its field-level resolvers.
+ */
+export type TypeSystem = {
+  [fqn in string]: TypeSpec
+};
+
+export namespace TypeSystem {
```
Comment (harrysolovay) [inline packages/punchcard/lib/appsync/type-system.ts:18]: I don't see a record type for GraphQL input objects. This specialized type needs to ban circularity, so no pointer shapes... I believe it also needs to be flat (only containing primitives). Could be wrong on that last statement.

Comment (harrysolovay) [reply packages/punchcard/lib/appsync/type-system.ts:18]: Actually, this doesn't apply to AppSync, given that the creation of resolvers is happening through the virtual VTL.

File: packages/punchcard/lib/appsync/util/time.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
+import { integer, string, timestamp } from '@punchcard/shape';
+import { VExpression } from '../expression';
+import { set } from '../statement';
+import { VTL } from '../vtl';
+import { VInteger, VObject, VString, VTimestamp } from '../vtl-object';
+
+export class TimeUtil {
```
Comment (harrysolovay) [inline packages/punchcard/lib/appsync/util/time.ts:7]: This approach to codegen is magnificent.

File: packages/punchcard/lib/appsync/vtl-object.ts [hunk 9]
```diff
@@ -0,0 +1,208 @@
+import { AnyShape, BinaryShape, bool, integer, NeverShape, NothingShape, ShapeGuards, ShapeVisitor, UnknownShape, Value } from '@punchcard/shape';
```
Comment (harrysolovay) [inline packages/punchcard/lib/appsync/vtl-object.ts:1]: You're creating so many bindings when `import * as` exists. Might I ask why?

File: packages/punchcard/lib/appsync/vtl-object.ts [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+import { ArrayShape, BoolShape, DynamicShape, IntegerShape, MapShape, NumberShape, Pointer, RecordShape, SetShape, Shape, StringShape, TimestampShape } from '@punchcard/shape';
+import { string, Trait } from '@punchcard/shape';
+import { FunctionArgs, FunctionShape } from '@punchcard/shape/lib/function';
+import { VExpression } from './expression';
+
+const type = Symbol.for('GraphQL.Type');
+const expr = Symbol.for('GraphQL.Expression');
+
+export class VObject<T extends Shape = Shape> {
```
Comment (harrysolovay) [inline packages/punchcard/lib/appsync/vtl-object.ts:10]: Why is the fallback type needed here?... won't it be inferred as the constructor's 1st arg?... or would removing it result in a 1-generic-expected TS error?

File: packages/punchcard/lib/appsync/vtl-object.ts [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+  [Trait.Data]: {
+    graphqlType: 'ID'
+  }
+} = {
+  [Trait.Data]: {
+    graphqlType: 'ID'
+  }
+};
+
+export const ID = string.apply(IDTrait);
```
Comment (harrysolovay) [inline packages/punchcard/lib/appsync/vtl-object.ts:90]: When extending this for other GraphQL use cases, it'll be important to enable non-string IDs. What other approaches could you take to creating the ID shape?

File: packages/punchcard/lib/appsync/vtl-object.ts [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+export class VNever extends VObject<NeverShape> {}
+export class VBinary extends VObject<BinaryShape> {}
+
+export class VBool extends VObject<BoolShape> {
+  public static not(a: VBool): VBool {
+    return new VBool(bool, VObject.exprOf(a).prepend('!'));
+  }
+}
+
+export class VString extends VObject<StringShape> {
```
Comment (harrysolovay) [inline packages/punchcard/lib/appsync/vtl-object.ts:106]: Can you please describe why you're taking this approach?

File: packages/punchcard/lib/appsync/vtl-object.ts [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+export class VTimestamp extends VObject<TimestampShape> {}
+
+export class VList<T extends VObject = VObject> extends VObject<ArrayShape<VObject.TypeOf<T>>> {
+  constructor(shape: ArrayShape<VObject.TypeOf<T>>, expression: VExpression) {
+    super(shape, expression);
+  }
+}
+
+
+export class VSet<T extends VObject = VObject> extends VObject<SetShape<VObject.TypeOf<T>>> {
```
Comment (harrysolovay) [inline packages/punchcard/lib/appsync/vtl-object.ts:133]: At first glance, this has me thing about how there is no `Set` type in GraphQL. What'm I missing here?

File: packages/punchcard/lib/appsync/vtl-object.ts [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+  public binaryShape(shape: BinaryShape, expr: VExpression): VBinary {
+    return new VBinary(shape, expr);
+  }
+  public boolShape(shape: BoolShape, expr: VExpression): VBool {
+    return new VBool(shape, expr);
+  }
+  public recordShape(shape: RecordShape<any>, expr: VExpression): VRecord {
+    return new VRecord(shape, expr);
+  }
+  public dynamicShape(shape: DynamicShape<any>, expr: VExpression): VAny | VUnknown {
```
Comment (harrysolovay) [inline packages/punchcard/lib/appsync/vtl-object.ts:178]: Similar to the comment above... will this shape ever be encountered by the visitor?

File: packages/punchcard/lib/appsync/vtl.ts [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+import { VExpression } from './expression';
+import { set, Statement } from './statement';
+import { VFloat, Visitor, VObject, VString } from './vtl-object';
+
+/**
+ * Represents a Velocity Template program.
+ */
+export type VTL<T> = Generator<Statement<VObject | void>, T>;
+
+declare module './vtl-object' {
```
Comment (harrysolovay) [inline packages/punchcard/lib/appsync/vtl.ts:12]: Why do you take this module augmentation approach? Can these `VObject` methods be defined within the `vtl-object` module instead?

File: packages/punchcard/lib/appsync/vtl.ts [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+};
+
+/**
+ * Type of an ExpressionTemplate factory.
+ *
+ * ```ts
+ * GraphQL.string`${mustBeAGraphQLType}`;
+ * ```
+ */
+export type ExpressionTemplate<T extends Shape> = <Args extends (VObject)[]>(template: TemplateStringsArray,...args: Args) => VTL<VObject.Of<T>>;
```
Comment (harrysolovay) [inline packages/punchcard/lib/appsync/vtl.ts:32]: Add a space after `TemplateStringsArray,`... we gotta get Prettier set up in this repo ;)

File: packages/punchcard/lib/appsync/vtl.ts [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+ * ```
+ *
+ * Translates to:
+ * ```
+ * #set($var1 = "hello")
+ * ```
+ *
+ * @param type what type to treat the assigned variable
+ */
+export function vtl<T extends Shape>(type: T): ExpressionTemplate<T> {
```
Comment (harrysolovay) [inline packages/punchcard/lib/appsync/vtl.ts:50]: Where is this used? Are tagged templates truly a good fit?

File: packages/punchcard/lib/appsync/vtl-object.ts (hunk 14 above)
Comment (sam-goodwin) [reply packages/punchcard/lib/appsync/vtl-object.ts:178]: While these types are not valid GraphQL schema types, the, we still need to interact with services like DynamoDB, SQS, SNS, Lambda which all have "VTL Representations".  For example, I treat sets as an array in JSON, but have to convert to a specific set type when serializing to/from DynamoDB.

We should be able to make use of the resolver utilities to implement some behavior for any. We could, for example, look for a "type tag" and cast to a well known type, or implement some generic serialization logic within VTL.

File: packages/punchcard/lib/appsync/vtl.ts (hunk 15 above)
Comment (sam-goodwin) [reply packages/punchcard/lib/appsync/vtl.ts:12]: Was getting circular dependencies. Been meaning to come back and fix this, it just hasn't been a priority. Will fix.

File: packages/punchcard/lib/appsync/vtl.ts (hunk 16 above)
Comment (sam-goodwin) [reply packages/punchcard/lib/appsync/vtl.ts:32]: Prove to me prettier isn't ugly, and I'm so down. :P

File: packages/punchcard/lib/appsync/vtl.ts (hunk 17 above)
Comment (sam-goodwin) [reply packages/punchcard/lib/appsync/vtl.ts:50]: I think it's an awesome fit. A bit like how styled components lets you write and generate CSS, this lets you write and generate and interact with VTL. I capture the template arguments (which are required to be instances of VObject) and then incorporate them into the VTL template safely.

The `type` argument tells the interpreter what type this data is, so the returned result is an object of that type.

```ts
const sam: VString = ;

const msg = yield* vtl(string)`hello ${sam}`

msg.isEmpty(); // can interact with the msg result like a String in Java
```

Might result in some VTL like:
```
#set($var1 = "sam")
#set($var2 = "hello ${var1}"
```

Note how it generates variables and sets them with values. That's how I perform computations in VTL, each statement sets a variable. I then can use those variables to generate the request/response payloads.

Comment (harrysolovay) [reply packages/punchcard/lib/appsync/vtl.ts:51]: Very very cool!

**Pull request merged**